### PR TITLE
Decompile weapon 21 (food) func_ptr_80170004

### PR DESCRIPTION
**Pull request by bismurphy**
Unfortunately, this required an evil change of the return type, even though it doesn't return. And since it's one of many weapons that share the functions, they all had to change.

Otherwise, nice to have the food figured out (this is the last food function), since food is a somewhat major part of this game.

### CODE REVIEW

**Review by sozud — APPROVED** [review]
Looks good to me, thinking Xeeynamo should look at this one too due to the API changes

**Comment by bismurphy** [comment]
Sounds good. For context, this is the same kind of change as was applied in https://github.com/Xeeynamo/sotn-decomp/pull/1128. (Wow, there have been 140 PRs since then! Crazy!)

File: src/weapon/w_021.c [hunk 1]
```diff
@@ -415,7 +415,114 @@ void EntityWeaponAttack(Entity* self) {
...
+        self->animCurFrame = upperParams - 0x1C;
+        self->posY.i.hi -= 4;
+        SetSpeedX((rand() & 0x3FFF) + FIX(0.875));
+        self->velocityY = FIX(-2);
+        DestroyEntityWeapon(1);
+        self->facingLeft = 0;
+        self->hitboxWidth = 8;
+        self->hitboxHeight = 8;
+        self->step++;
+        return;
```
Comment (Xeeynamo) [inline src/weapon/w_021.c:452]:
```suggestion
        break;
```

File: src/weapon/w_021.c [hunk 2]
```diff
@@ -415,7 +415,114 @@ void EntityWeaponAttack(Entity* self) {
...
+                g_api.FreePrimitives(g_unkGraphicsStruct.BottomCornerTextPrims);
+                g_unkGraphicsStruct.BottomCornerTextTimer = 0;
+            }
+            g_api.PlaySfx(0x67C);
+            temp_s1 = g_api.equipDefs[upperParams].name;
+            g_api.AddToInventory(upperParams, EQUIP_HAND);
+            func_97000_8017AB54(temp_s1, 1);
+            DestroyEntity(self);
+        }
+    }
```
Comment (Xeeynamo) [inline src/weapon/w_021.c:524]:
```suggestion
        break;
    }
```
let's add a `break` there, it makes easier to read the end of a `switch` statement

File: src/weapon/w_021.c [hunk 3]
```diff
@@ -415,7 +415,114 @@ void EntityWeaponAttack(Entity* self) {
...
+    const char* temp_s1;
+    s16 xVar;
+    s16 yVar;
+
+    upperParams = (self->params >> 8) & 0x7F;
+
+    switch (self->step) {
+    case 0:
+        SetSpriteBank1(D_97000_8017A040);
+        self->animSet = -0x7FF0;
```
Comment (Xeeynamo) [inline src/weapon/w_021.c:432]: This should be `ANIM_OVL(0x10)` I think

**Review by Xeeynamo — APPROVED** [review]
Cursed match but it's okay if it is required. I left a few minor :nail_care: , other than that it looks good

**Pull request merged**